mava/evaluator.py

```diff
@@ -12,329 +12,140 @@
-
-    Returns:
-        Tuple[EvalFn, EvalFn]: A tuple of two evaluation functions:
-        one for use during training and one for absolute metrics.
-
-    Raises:
-        AssertionError: If `use_recurrent_net` is True but `scanned_rnn` is not provided.
+        env: an environment that conforms to the mava environment spec.
+        act_fn: a function that takes in params, timestep, key and optionally a state
+                and returns actions and optionally a state (see `EvalActFn`).
+        config: the systems config.
+        eval_episodes: how many episodes to evaluate the agents. At least this many episodes
+                will be done, however more may be done if `config.arch.num_envs` does
+                not evenly divide `eval_episodes`
     """
-    # Check if win rate is required for evaluation.
-    log_win_rate = config.env.log_win_rate
-    # Vmap it over number of agents and create evaluator_fn.
-    if use_recurrent_net:
-        assert scanned_rnn is not None
-        evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-        )
-        absolute_metric_evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-            10,
-        )
-    else:
-        evaluator = get_ff_evaluator_fn(
-            eval_env, network_apply_fn, config, log_win_rate  # type: ignore
-        )
-        absolute_metric_evaluator = get_ff_evaluator_fn(
-            eval_env, network_apply_fn, config, log_win_rate, 10  # type: ignore
-        )
-
-    evaluator = jax.pmap(evaluator, axis_name="device")
-    absolute_metric_evaluator = jax.pmap(absolute_metric_evaluator, axis_name="device")
-
-    return evaluator, absolute_metric_evaluator
+
+    def eval_fn(params: FrozenDict, key: PRNGKey, init_act_state: ActorState) -> Dict[str, Array]:
+        def _env_step(eval_state: _EvalEnvStepState, _: Any) -> Tuple[_EvalEnvStepState, TimeStep]:
+            """Performs a single environment step"""
+            env_state, ts, key, actor_state = eval_state
+
+            key, act_key = jax.random.split(key)
+            action, actor_state = act_fn(params, ts, act_key, **actor_state)
+            env_state, ts = jax.vmap(env.step)(env_state, action)
+
+            return (env_state, ts, key, actor_state), ts
+
+        def _episode(key: PRNGKey, _: Any) -> Tuple[PRNGKey, Metrics]:
+            """Simulates `config.arch.num_envs` episodes."""
+            key, reset_key = jax.random.split(key)
+            reset_keys = jax.random.split(reset_key, config.arch.num_envs)
+            env_state, ts = jax.vmap(env.reset)(reset_keys)
+
+            step_state = env_state, ts, key, init_act_state
+            _, timesteps = jax.lax.scan(_env_step, step_state, jnp.arange(env.time_limit))
```

Comment:
how do you make sure the env didn't do an autorest or it gets wrong info after done equal true?

Reply:
Great question, I let the env run for `time_limit` even if it is passed done (I don't see a way around with jax), but I make sure I only take the metrics from the final step in the first episode. This line gets the index: `done_idx = jnp.argmax(timesteps.last(), axis=0)` and this uses that index on the metrics: `metrics = jax.tree_map(lambda m: m[done_idx, jnp.arange(config.arch.num_envs)], metrics)`.

Reply:
oh interesting, yeah the argmax will take the first time we have done=1, yeah nothing to worry about especially that all our envs have timelimit